🔧  add : 체험 프로그램 생성 API 파일 업로드 로직 추가

j-hann opened the pull request:
🟪 `http` 디렉토리 & 내부 `.http` 파일 `resources` 디렉토리로 이동
</br>

🟪 ProgramContoller `createProgram` 메서드 파일 업로드 코드 추가 및 수정
   - Swagger `security` 토큰 명칭 `bearerAuth`로 변경
   - `@RequestPart`를 사용해 `JSON` 문자열과 `file` 리스트를 각각 분리해서 받도록 구성
     - 클라이언트가 `JSON` 데이터와 `file`을 동시에 전송할 수 있도록 설계
   - `JSON` 문자열 `ProgramCreateRequestDto`로 변환하기 위해 `ObjectMapper.readValue()` 사용해 수동 역직렬화 처리
   - `ObjectMapper`는 Spring Boot가 자동 구성한 Jackson 설정 적용받기 위해 직접 생성하지 않고 생성자 주입으로 사용
      -  `JavaTimeModule`, `WRITE_DATES_AS_TIMESTAMPS=false`, `jackson-datatype-jsr310`  등
</br>

🟪 ProgramService  `createProgram` 메서드 파일 업로드 로직 추가
   - 업로드 파일 유효성 검사 로직 추가 `최대 10개, JPG, JPEG, PNG 타입만 허용`
   - `S3Uploader`를 통해 S3 `programs/` 경로에 파일 업로드
   - `File` Entity 생성 & 저장
      - 원본 파일 이름 & S3 업로드 된 파일 url
   -  `FileMapping` Entity 생성 & 저장
      - 업로드된 파일 `PROGRAM` 카테고리로 매핑
      - `referenceId` 필드에 `Program id` 저장하여 연관 정보 설정
</br>

🟪 ProgramCreateRequestDto 필드 수정
   - `title`, `description`, `curriculumDescription`, `location` 필드 @NotBlank로 재변경
</br>

🟪 ErrorCode 추가
</br>

🟪 `FilePurpose` Enum 값 추가
  - `THUMBNAIL`, `DETAIL_IMAGE`
</br>

🟪 ProgramController 
  - `@RequestPram`으로 `filePurposes` 데이터 받아서 처리
  - 전달 받은 문자열 리스트를 `FilePurpose` Enum 타입의 값들로 변환하여 `ProgramService`에 전달

🟪 ProgramService 
   - `files` &` filePurposes` 인덱스 기반 매핑 처리
   - 파일 유효성 검증 메서드 분리
      - `파일 타입`, `파일 개수`, `파일 목적`, `썸네일 개수` 검증

j-hann committed:
🔧  add : 체험 프로그램 생성 API에 File 업로드 로직 추가

- http 디렉토리 & 내부 파일 resources로 이동, - ProgramContoller 파일 업로드 위해 multipart/form-data 방식으로 변경, - ObjectMapper로 JSON 문자열 수동 역직렬화 처리, - Swagger security 토큰 명칭 변경

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/program/service/ProgramService.java:
```diff
@@ -144,8 +154,41 @@ public ProgramCreateResponseDto createProgram(long memberId, @Valid ProgramCreat
+			}
+
+			for(MultipartFile multipartFile : files){
+
```
`filePurpose` `null 허용`하는거에 대해서 계속 생각해보다가 문득 든 생각이예요!!☺️(아닌것 같으면 넘기셔도 됩니다.)

저희 UXUI 디자인 보면 체험프로그램 리스트에 대표이미지가 하나 들어가고, 상세페이지 들어가면 이미지가 여러개 보이는 구조인데, 이럴 경우 `filePurpose`에 썸네일, 상세이미지 이런식으로 구분하게 되면 리스트 조회시에 썸네일 이미지만 불러오기 편할 것 같아요!

그리고 이후 서비스 운영을 생각해봤을때 상세이미지는 두고 썸네일만 바꾼다던가, 상세 이미지들에 정렬기능을 넣으면 사진 순서 변경도 쉽지 않을까? 라는 생각을 해봤습니당😎

j-hann replied:
좋은 방법 같아요! 반영해서 수정했습니다~🥰

jiyoon0000 replied:
짱이예요!!!!👍

jiyoon0000 commented on src/main/resources/http/Auth.http:
auth....도 소문자로 해주시면 안돼요옹?🥺

j-hann replied:
ㅋㅋㅋㅋㅋㅋ네넵 수정했어요!

j-hann committed:
🔧  add : 체험 프로그램 생성 API - FilePurpose(파일 목적) 구분 로직 추가

- FilePurpose Enum 값 추가, - ProgramController @RequestPram으로 filePurposes 데이터 받아서 처리, - ProgramService files & filePurposes 인덱스 기반 매핑 처리, - 파일 유효성 검증 로직 추가, - ErrorCode 추가
